### ISSUE TITLE
What will happen if I change the backbone network into MobileNet-v2 with FPN? #27

### ISSUE THREAD

**Issue by nivosco** [dbolya/yolact#214]

**Comment by nivosco** [comment]
Hi,
Just wanted to share with you my code for adding Mobilenet-V1 to the available backbones.

backbone.py:
`class MobileNetBackbone(nn.Module):
    def __init__(self, extra_args=[]):
        super().__init__()
        self.layers = nn.ModuleList()
        self.channels = []
        self.num_base_layers = 4
        self.layer_to_block = {0: '_preconv.', 1: 'layers.0.', 2: 'layers.0.', 3: 'layers.1.', 4: 'layers.1.', 5: 'layers.2.', 6: 'layers.2.'}

        self._preconv = self.conv_bn(3, 32, 2)
        self._make_layer(2, [(32, 64, 1), (64, 128, 2)])
        self._make_layer(2, [(128, 128, 1), (128, 256, 2)])
        self._make_layer(2, [(256, 256, 1), (256, 512, 2)])
        self._make_layer(6, [(512, 512, 1), (512, 512, 1), (512, 512, 1), (512, 512, 1), (512, 512, 1), (512, 1024, 2)])

        self.backbone_modules = [m for m in self.modules() if isinstance(m, nn.Conv2d)]

    def _make_layer(self, i, cfg):
        """ Here one layer means a string of n blocks. """
        layer_list = []

        # The downsample layer
        for idx in range(i):
            layer_list.append(
                self.conv_dw(cfg[idx][0], cfg[idx][1], cfg[idx][2])
            )
        self.channels.append(cfg[-1][1])
        self.layers.append(nn.Sequential(*layer_list))

    def conv_bn(self, inp, oup, stride):
        return nn.Sequential(
            nn.Conv2d(inp, oup, 3, stride, 1, bias=False),
            nn.BatchNorm2d(oup),
            nn.ReLU(inplace=True)
        )
    def conv_dw(self, inp, oup, stride):
        return nn.Sequential(
            nn.Conv2d(inp, inp, 3, stride, 1, groups=inp, bias=False),
            nn.BatchNorm2d(inp),
            nn.ReLU(inplace=True),
            nn.Conv2d(inp, oup, 1, 1, 0, bias=False),
            nn.BatchNorm2d(oup),
            nn.ReLU(inplace=True),
        )

    def transform_key(self, k):
        vals = k.split('.')
        if 'base_net' in k:
            block = self.layer_to_block.get(int(vals[1]), 'layers.3.')
            return block + vals[2] + '.' + vals[3]
        return k

    def init_backbone(self, path):
        state_dict = torch.load(path)
        state_dict = OrderedDict([(self.transform_key(k), v) for k,v in state_dict.items()])
        """ Initializes the backbone weights for training. """
        self.load_state_dict(state_dict, strict=False)

    def forward(self, x):
        x = self._preconv(x)
        outs = []
        for layer in self.layers:
            x = layer(x)
            outs.append(x)
        return tuple(outs)

    def add_layer(self, conv_channels=1024, downsample=2):
        """ Add a downsample layer to the backbone as per what SSD does. """
        self._make_layer(1, [(conv_channels, conv_channels, downsample)])`

data/config.py
`mobilenet_v1_backbone = resnet50_backbone.copy({
    'name': 'MobileNet_v1',
    'path': 'mobilenet-v1-ssd-mp-0_675.pth',
    'type': MobileNetBackbone,
})
yolact_mobilenet_v1_config = yolact_base_config.copy({
    'name': 'yolact_mobilenet_v1',

    'backbone': mobilenet_v1_backbone.copy({
        'pred_scales': yolact_base_config.backbone.pred_scales,
        'pred_aspect_ratios': yolact_base_config.backbone.pred_aspect_ratios,
        'selected_layers': list(range(1, 4)),

        'use_pixel_scales': True,
        'preapply_sqrt': False,
        'use_square_anchors': True, # This is for backward compatability with a bug
    }),
})`

I got ~18mAP. Do you have any ideas of how to improve that result?

**Comment by dbolya** [comment]
Thanks for the code! Do you know what the performance hit for other methods using mobilenet vs resnet101 is? If we know that, we can work out if 18 mAP is a reasonable figure or if there are some optimizations to be made.

**Comment by zhenyouwei** [comment]
@nivosco   Thank U, I used the code you shared,but the APbox I got is only 0.01,Can you tell me what may be the problem?

**Comment by zhenyouwei** [comment]
@nivosco   Thank U, I used the code you shared,but the APbox I got is only 0.01,Can you tell me what may be the problem?

**Comment by KennyChuang888** [comment]
I guess it's because you're choosing layers, because you're choosing layers 1 to 4, you're probably missing high semantic information, and if you're choosing layers that are evenly distributed, it's theoretically better

**Comment by jonguo111** [comment]
> I guess it's because you're choosing layers, because you're choosing layers 1 to 4, you're probably missing high semantic information, and if you're choosing layers that are evenly distributed, it's theoretically better

Hi. So for the mobienetv2 which layers are more suitable for yolact training? Do you have some suggestions? Thank you so much

**Comment by azade-a** [comment]
@nivosco  Hello. First of all thanks for sharing your implementation.
I used your mobilenetv1 backbone code but I didn't get good results. My custom dataset has only one class with medium and large objects.
I don't know which layers are more suitable for my dataset. Do you have some suggestions?
 Thank you so much

**Comment by nivosco** [comment]
@azade-a  I found it quite important to make sure pretrained weights of the backbone are loaded properly.

**Comment by azade-a** [comment]
> 
> 
> @azade-a I found it quite important to make sure pre-trained weights of the backbone are loaded properly.

where are these weights loaded? 
How can I find the best pred scales and ratios ? 
Many thanks in advance
